File: Exam/bst.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BST:
    def __init__(self):
        self.root = None

    # Insert a node
    def insert(self, root, data):
        if root is None:
            return Node(data)
        if data < root.data:
            root.left = self.insert(root.left, data)
        elif data > root.data:
            root.right = self.insert(root.right, data)
        return root

    # Infix (inorder) traversal
    def inorder(self, root):
        if root:
            self.inorder(root.left)
            print(root.data, end=' ')
            self.inorder(root.right)

    # Prefix (preorder)
    def preorder(self, root):
        if root:
            print(root.data, end=' ')
            self.preorder(root.left)
            self.preorder(root.right)

    # Postfix (postorder)
    def postorder(self, root):
        if root:
            self.postorder(root.left)
            self.postorder(root.right)
            print(root.data, end=' ')

    # Count nodes
    def countNodes(self, root):
        if root is None:
            return 0
        return 1 + self.countNodes(root.left) + self.countNodes(root.right)

    # Height
    def height(self, root):
        if root is None:
            return 0
        left_height = self.height(root.left)
        right_height = self.height(root.right)
        return 1 + max(left_height, right_height)

    # Count leaf nodes
    def countLeaf(self, root):
        if root is None:
            return 0
        if root.left is None and root.right is None:
            return 1
        return self.countLeaf(root.left) + self.countLeaf(root.right)

    # Count internal nodes
    def countInternal(self, root):
        if root is None or (root.left is None and root.right is None):
            return 0
        return 1 + self.countInternal(root.left) + self.countInternal(root.right)

    # Search a node
    def search(self, root, key):
        if root is None or root.data == key:
            return root
        if key < root.data:
            return self.search(root.left, key)
        return self.search(root.right, key)

    # Find minimum value
    def findMin(self, root):
        current = root
        while current.left:
            current = current.left
        return current.data

    # Find maximum value
    def findMax(self, root):
        current = root
        while current.right:
            current = current.right
        return current.data

    # Count NULL pointers (possible children spots that are empty)
    def countNullPointers(self, root):
        if root is None:
            return 1
        return self.countNullPointers(root.left) + self.countNullPointers(root.right)

    # Check if complete binary tree helper
    def isCompleteUtil(self, root, index, number_nodes):
        if root is None:
            return True
        if index >= number_nodes:
            return False
        return (self.isCompleteUtil(root.left, 2 * index + 1, number_nodes) and
                self.isCompleteUtil(root.right, 2 * index + 2, number_nodes))

    def isComplete(self, root):
        node_count = self.countNodes(root)
        return self.isCompleteUtil(root, 0, node_count)

    # Find node height
    def getBalanceFactor(self, node):
        if not node:
            return 0
        return self.height(node.left) - self.height(node.right)

    # Delete a node
    def delete(self, root, key):
        if root is None:
            return root
        if key < root.data:
            root.left = self.delete(root.left, key)
        elif key > root.data:
            root.right = self.delete(root.right, key)
        else:
            # node with one or no child
            if root.left is None:
                temp = root.right
                root = None
                return temp
            elif root.right is None:
                temp = root.left
                root = None
                return temp

            # node with two children
            temp_val = self.findMin(root.right)
            root.data = temp_val
            root.right = self.delete(root.right, temp_val)
        return root

# Example usage
if __name__ == "__main__":
    bst = BST()
    bst.root = None
    while True:
        print("\nMenu:")
        print("1. Insert")
        print("2. Inorder traversal")
        print("3. Preorder traversal")
        print("4. Postorder traversal")
        print("5. Count nodes")
        print("6. Height")
        print("7. Count leaf nodes")
        print("8. Count internal nodes")
        print("9. Search")
        print("10. Check if complete binary tree")
        print("11. Delete node")
        print("12. Get balance factor of root")
        print("13. Find min value")
        print("14. Find max value")
        print("15. Count null pointers")
        print("0. Exit")
        choice = int(input("Enter your choice: "))

        if choice == 1:
            val = int(input("Enter value to insert: "))
            bst.root = bst.insert(bst.root, val)
        elif choice == 2:
            print("Inorder traversal: ", end="")
            bst.inorder(bst.root)
            print()
        elif choice == 3:
            print("Preorder traversal: ", end="")
            bst.preorder(bst.root)
            print()
        elif choice == 4:
            print("Postorder traversal: ", end="")
            bst.postorder(bst.root)
            print()
        elif choice == 5:
            print("Total nodes: ", bst.countNodes(bst.root))
        elif choice == 6:
            print("Height: ", bst.height(bst.root))
        elif choice == 7:
            print("Leaf nodes: ", bst.countLeaf(bst.root))
        elif choice == 8:
            print("Internal nodes: ", bst.countInternal(bst.root))
        elif choice == 9:
            k = int(input("Enter value to search: "))
            result = bst.search(bst.root, k)
            if result:
                print(f"{k} found in the tree.")
            else:
                print(f"{k} not found in the tree.")
        elif choice == 10:
            print("Complete Binary Tree?" , "Yes" if bst.isComplete(bst.root) else "No")
        elif choice == 11:
            val = int(input("Enter node value to delete: "))
            bst.root = bst.delete(bst.root, val)
            print(f"Node {val} deleted.")
        elif choice == 12:
            print("Balance factor at root:", bst.getBalanceFactor(bst.root))
        elif choice == 13:
            if bst.root:
                print("Minimum value:", bst.findMin(bst.root))
            else:
                print("Tree is empty")
        elif choice == 14:
            if bst.root:
                print("Maximum value:", bst.findMax(bst.root))
            else:
                print("Tree is empty")
        elif choice == 15:
            print("Number of NULL pointers:", bst.countNullPointers(bst.root))
        elif choice == 0:
            print("Exiting...")
            break
        else:
            print("Invalid choice")

File: Exam/bst.c
#include <stdio.h>
#include <stdlib.h>

struct node {
    int data;
    struct node *left, *right;
};

// Create a new node
struct node* createNode(int data) {
    struct node* n = (struct node*) malloc(sizeof(struct node));
    n->data = data;
    n->left = n->right = NULL;
    return n;
}

// Insert a node (BST property)
struct node* insert(struct node* root, int data) {
    if (!root)
        return createNode(data);
    if (data < root->data)
        root->left = insert(root->left, data);
    else if (data > root->data)
        root->right = insert(root->right, data);
    return root;
}

// Infix (inorder)
void inorder(struct node* root) {
    if (root) {
        inorder(root->left);
        printf("%d ", root->data);
        inorder(root->right);
    }
}

// Prefix (preorder)
void preorder(struct node* root) {
    if (root) {
        printf("%d ", root->data);
        preorder(root->left);
        preorder(root->right);
    }
}

// Postfix (postorder)
void postorder(struct node* root) {
    if (root) {
        postorder(root->left);
        postorder(root->right);
        printf("%d ", root->data);
    }
}

// Count nodes
int countNodes(struct node* root) {
    if (!root) return 0;
    return 1 + countNodes(root->left) + countNodes(root->right);
}

// Height of tree
int height(struct node* root) {
    if (!root) return 0;
    int lh = height(root->left);
    int rh = height(root->right);
    return 1 + (lh > rh ? lh : rh);
}

// Count leaf nodes
int countLeaf(struct node* root) {
    if (!root) return 0;
    if (!root->left && !root->right)
        return 1;
    return countLeaf(root->left) + countLeaf(root->right);
}

// Count internal nodes
int countInternal(struct node* root) {
    if (!root || (!root->left && !root->right))
        return 0;
    return 1 + countInternal(root->left) + countInternal(root->right);
}

// Search for a data value
struct node* search(struct node* root, int key) {
    if (!root) return NULL;
    if (root->data == key) return root;
    else if (key < root->data) return search(root->left, key);
    else return search(root->right, key);
}

// Find minimum value
int findMin(struct node* root) {
    if (!root) return -1;
    while (root->left) root = root->left;
    return root->data;
}

// Find maximum value
int findMax(struct node* root) {
    if (!root) return -1;
    while (root->right) root = root->right;
    return root->data;
}

// Count NULL pointers
int countNullPointers(struct node* root) {
    if (!root) return 1;
    return countNullPointers(root->left) + countNullPointers(root->right);
}

// Balance factor of a node (absolute difference of heights)
int balanceFactor(struct node* root) {
    if (!root) return 0;
    int lh = height(root->left);
    int rh = height(root->right);
    return (lh > rh) ? lh - rh : rh - lh;
}

// Check if complete binary tree
int isComplete(struct node* root, int index, int node_count) {
    if (!root) return 1;
    if (index >= node_count) return 0;
    return isComplete(root->left, 2*index+1, node_count) &&
           isComplete(root->right, 2*index+2, node_count);
}

// Deletion in BST
struct node* deleteNode(struct node* root, int key) {
    if (!root) return NULL;
    if (key < root->data)
        root->left = deleteNode(root->left, key);
    else if (key > root->data)
        root->right = deleteNode(root->right, key);
    else {
        // Found node
        if (!root->left) {
            struct node* temp = root->right;
            free(root);
            return temp;
        }
        else if (!root->right) {
            struct node* temp = root->left;
            free(root);
            return temp;
        }
        // Node with two children
        int minVal = findMin(root->right);
        root->data = minVal;
        root->right = deleteNode(root->right, minVal);
    }
    return root;
}

// Main program
int main() {
    struct node* root = NULL;
    int choice, val;

    while (1) {
        printf("\nMenu:\n");
        printf("1. Insert Node\n");
        printf("2. Infix (Inorder)\n");
        printf("3. Prefix (Preorder)\n");
        printf("4. Postfix (Postorder)\n");
        printf("5. Count Nodes\n");
        printf("6. Height of Tree\n");
        printf("7. Count Leaf Nodes\n");
        printf("8. Count Internal Nodes\n");
        printf("9. Search\n");
        printf("10. Complete Binary Tree Check\n");
        printf("11. Delete Node\n");
        printf("12. Balance Factor (at root)\n");
        printf("13. Find Minimum\n");
        printf("14. Find Maximum\n");
        printf("15. Count NULL Pointers\n");
        printf("0. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch(choice) {
            case 1:
                printf("Enter value to insert: ");
                scanf("%d", &val);
                root = insert(root, val);
                break;
            case 2:
                printf("Inorder: ");
                inorder(root);
                printf("\n");
                break;
            case 3:
                printf("Preorder: ");
                preorder(root);
                printf("\n");
                break;
            case 4:
                printf("Postorder: ");
                postorder(root);
                printf("\n");
                break;
            case 5:
                printf("Node count: %d\n", countNodes(root));
                break;
            case 6:
                printf("Tree height: %d\n", height(root));
                break;
            case 7:
                printf("Leaf node count: %d\n", countLeaf(root));
                break;
            case 8:
                printf("Internal node count: %d\n", countInternal(root));
                break;
            case 9:
                printf("Enter value to search: ");
                scanf("%d", &val);
                if (search(root, val))
                    printf("Found\n");
                else
                    printf("Not Found\n");
                break;
            case 10: {
                int nodes = countNodes(root);
                if (isComplete(root, 0, nodes))
                    printf("Tree is complete\n");
                else
                    printf("Tree is NOT complete\n");
                break;}
            case 11:
                printf("Enter node to delete: ");
                scanf("%d", &val);
                root = deleteNode(root, val);
                printf("Node deleted if existed.\n");
                break;
            case 12:
                printf("Balance factor (root): %d\n", balanceFactor(root));
                break;
            case 13:
                printf("Minimum value: %d\n", findMin(root));
                break;
            case 14:
                printf("Maximum value: %d\n", findMax(root));
                break;
            case 15:
                printf("Number of NULL pointers: %d\n", countNullPointers(root));
                break;
            case 0:
                printf("Exiting...\n");
                return 0;
            default:
                printf("Invalid option\n");
        }
    }
    return 0;
}
